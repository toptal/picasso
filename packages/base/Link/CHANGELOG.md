# @toptal/picasso-link

## 3.0.2

### Patch Changes

- [#4547](https://github.com/toptal/picasso/pull/4547) [`2984ddf`](https://github.com/toptal/picasso/commit/2984ddf167dfc3a08ab7e978ef31ebfc69eae163) Thanks [@OleksandrNechai](https://github.com/OleksandrNechai)!

### Link

- fix: protect Link from unexpected input

## 3.0.1

### Patch Changes

- Updated dependencies [[`80407eb`](https://github.com/toptal/picasso/commit/80407eb734c69894ee6d2dadd3e773752fc43c5d)]:
  - @toptal/picasso-utils@2.0.0
  - @toptal/picasso-typography@4.0.1

## 3.0.0

### Major Changes

- [#4500](https://github.com/toptal/picasso/pull/4500) [`3ed8c02`](https://github.com/toptal/picasso/commit/3ed8c0271982a82dd9cdc6b967c63656afd3654f) Thanks [@ruslan-sed](https://github.com/ruslan-sed)!
- update version of `@toptal/picasso-tailwind-merge` peer dependency

### Patch Changes

- Updated dependencies [[`3ed8c02`](https://github.com/toptal/picasso/commit/3ed8c0271982a82dd9cdc6b967c63656afd3654f)]:
  - @toptal/picasso-typography@4.0.0

## 2.0.0

### Major Changes

- [#4345](https://github.com/toptal/picasso/pull/4345) [`6716e0b`](https://github.com/toptal/picasso/commit/6716e0bb3178a7f452f2c79ce56dd524e9bd8685) Thanks [@augustobmoura](https://github.com/augustobmoura)!

### Link

- migrated `Link` to not use MUI and only use Tailwind CSS instead

- [#4345](https://github.com/toptal/picasso/pull/4345) [`6716e0b`](https://github.com/toptal/picasso/commit/6716e0bb3178a7f452f2c79ce56dd524e9bd8685) Thanks [@augustobmoura](https://github.com/augustobmoura)!
- add @toptal/picasso-tailwind-merge as peer dependency, it now needs to be provided by the project consuming the package

### Patch Changes

- [#4345](https://github.com/toptal/picasso/pull/4345) [`6716e0b`](https://github.com/toptal/picasso/commit/6716e0bb3178a7f452f2c79ce56dd524e9bd8685) Thanks [@augustobmoura](https://github.com/augustobmoura)!

### Link

- update line-height to match the previous MUI implementation
- Updated dependencies [[`6716e0b`](https://github.com/toptal/picasso/commit/6716e0bb3178a7f452f2c79ce56dd524e9bd8685), [`6716e0b`](https://github.com/toptal/picasso/commit/6716e0bb3178a7f452f2c79ce56dd524e9bd8685), [`6716e0b`](https://github.com/toptal/picasso/commit/6716e0bb3178a7f452f2c79ce56dd524e9bd8685)]:
  - @toptal/picasso-typography@3.0.0

## 1.0.3

### Patch Changes

- Updated dependencies []:
  - @toptal/picasso-utils@1.0.3

## 1.0.2

### Patch Changes

- Updated dependencies []:
  - @toptal/picasso-utils@1.0.2

## 1.0.1

### Patch Changes

- [#4164](https://github.com/toptal/picasso/pull/4164) [`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30) Thanks [@mkrl](https://github.com/mkrl)!
- migrate to Picasso by Parts
  - picasso is now distributed as a set of independent packages with the main package `@toptal/picasso` now being a collection of re-exported packages
- Updated dependencies [[`84d579cee65c88aeb6ba58318c43d8deb5201b30`](https://github.com/toptal/picasso/commit/84d579cee65c88aeb6ba58318c43d8deb5201b30)]:
  - @toptal/picasso-utils@1.0.1
