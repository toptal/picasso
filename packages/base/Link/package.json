{
  "name": "@toptal/picasso-link",
  "version": "1.0.0",
  "description": "Toptal UI components library - Link",
  "publishConfig": {
    "access": "public",
    "directory": "dist-package"
  },
  "main": "index.js",
  "module": "index.js",
  "scripts": {
    "build:package": "cross-env NODE_ENV=production node ../../../bin/build.js --tsConfig=./tsconfig.build.json",
    "prepublishOnly": "if [ -d dist-package ]; then cp ./package.json ./dist-package/package.json; fi"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/toptal/picasso.git"
  },
  "author": "Toptal",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/toptal/picasso/issues"
  },
  "homepage": "https://github.com/toptal/picasso/tree/master/packages/picasso#readme",
  "dependencies": {
    "classnames": "^2.3.1"
  },
  "sideEffects": [
    "**/styles.ts",
    "**/styles.js"
  ],
  "peerDependencies": {
    "react": ">=16.12.0 < 19.0.0",
    "@material-ui/core": "4.12.4"
  }
}