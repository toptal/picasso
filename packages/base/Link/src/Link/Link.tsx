import type { ReactNode, AnchorHTMLAttributes, ElementType } from 'react'
import React, { forwardRef } from 'react'
import { twMerge } from '@toptal/picasso-tailwind-merge'
import type { BaseProps, OverridableComponent } from '@toptal/picasso-shared'
import { Typography } from '@toptal/picasso-typography'

const supportedColors = ['white', 'blue'] as const
const supportedVariants = ['action', 'anchor'] as const

type VariantType = (typeof supportedVariants)[number]
type ColorType = (typeof supportedColors)[number]

const sanitizeRel = (rel: string | undefined, target: string | undefined) => {
  if (target !== '_blank') {
    return rel
  }

  if (!rel) {
    return 'noopener'
  }

  const isRelSafe = rel.includes('noreferrer') || rel.includes('noopener')

  return isRelSafe ? rel : rel.concat(' noopener')
}

const COLOR_DISABLED_MAP: Record<
  ColorType,
  Record<VariantType, Record<'disabled' | 'normal', string>>
> = {
  blue: {
    action: {
      disabled: 'text-blue-500 no-underline opacity-50 hover:!no-underline',
      normal: 'text-blue-500 visited:text-purple-500 no-underline',
    },
    anchor: {
      disabled: 'text-gray-600 underline',
      normal: 'text-blue-500 visited:text-purple-500 no-underline',
    },
  },
  white: {
    action: {
      disabled: 'text-gray-600',
      normal: 'inherit',
    },
    anchor: {
      disabled: 'text-gray-600 visited:text-gray-500 underline',
      normal: 'inherit text-white visited:text-gray-500 underline',
    },
  },
}

export type Props = BaseProps &
  AnchorHTMLAttributes<HTMLAnchorElement> & {
    /** Content of the component */
    children?: ReactNode
    /** Destination the link points to */
    href?: string
    /** Callback invoked when component is clicked */
    onClick?: (event: React.MouseEvent<HTMLAnchorElement>) => void
    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     */
    as?: ElementType<React.HTMLAttributes<HTMLElement>>
    /** Either it's a regular hyperlink or an _action_ */
    variant?: VariantType
    /** Controls color of the link */
    color?: ColorType
    /** Indicates the order of receiving focus. If not set will not receive focus. */
    tabIndex?: number
    /** Indicates that the user cannot interact with the Link or its children */
    disabled?: boolean

    visited?: boolean
    /**
     * If true, underline decoration never applies
     */
    noUnderline?: boolean
    'aria-disabled'?: boolean
  }

const defaultColor = 'blue'
const defaultVariant = 'anchor'

export const Link: OverridableComponent<Props> = forwardRef<
  HTMLAnchorElement,
  Props
>(function Link(props, ref) {
  const {
    href,
    onClick,
    children,
    className,
    color: inputColor = 'blue',
    style,
    as = 'a',
    variant: inputVariant = 'anchor',
    tabIndex,
    target,
    rel,
    disabled,
    visited = false,
    noUnderline,
    'aria-disabled': ariaDisabled,
    ...rest
  } = props
  const nativeHTMLAttributes = rest
  const sanitizedRel = sanitizeRel(rel, target)

  // When Link is used as={Link}, TypeScript can't ensure the input to the Link is compatible with its Props type.
  const color = supportedColors.includes(inputColor) ? inputColor : defaultColor
  const variant = supportedVariants.includes(inputVariant)
    ? inputVariant
    : defaultVariant

  return (
    <Typography
      {...nativeHTMLAttributes}
      ref={ref}
      as={as}
      // @ts-expect-error Typography is incompatible with href prop
      href={disabled ? undefined : href}
      target={disabled ? undefined : target}
      rel={sanitizedRel}
      onClick={disabled ? undefined : onClick}
      color='inherit'
      weight={variant === 'action' ? 'semibold' : 'inherit'}
      className={twMerge(
        'focus:outline-none hover:underline leading-[inherit]',
        COLOR_DISABLED_MAP[color][variant][disabled ? 'disabled' : 'normal'],
        disabled ? 'cursor-not-allowed' : '',
        noUnderline ? '!no-underline' : '',
        visited
          ? color === 'blue'
            ? 'visited text-purple-500'
            : 'visited text-gray-500'
          : '',
        className
      )}
      style={style}
      tabIndex={tabIndex}
      aria-disabled={disabled || ariaDisabled}
    >
      {children}
    </Typography>
  )
})

Link.defaultProps = {
  as: 'a',
  color: 'blue',
  variant: 'anchor',
  noUnderline: false,
}

Link.displayName = 'Link'

export default Link
