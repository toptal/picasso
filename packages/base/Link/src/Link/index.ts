import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Link'

export { default as Link } from './Link'
export type LinkProps = OmitInternalProps<Props>
