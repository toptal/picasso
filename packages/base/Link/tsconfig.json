{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": { "outDir": "dist-package" },
  "include": ["src"],
  "references": [
    { "path": "../../picasso-provider" },
    { "path": "../../picasso-tailwind" },
    { "path": "../../picasso-tailwind-merge" },
    { "path": "../Typography" },
    { "path": "../Utils" }
  ]
}
