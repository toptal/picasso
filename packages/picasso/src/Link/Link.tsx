import React, {
  forwardRef,
  ReactNode,
  ElementType,
  AnchorHTMLAttributes,
} from 'react'
import { Link as MUILink } from '@material-ui/core'
import { Theme, makeStyles } from '@material-ui/core/styles'
import cx from 'classnames'
import { BaseProps, OverridableComponent } from '@toptal/picasso-shared'

import styles from './styles'

type VariantType = 'action' | 'anchor'
type ColorType = 'white' | 'blue'

const useStyles = makeStyles<Theme>(styles, { name: 'PicassoLink' })

const sanitizeRel = (rel: string | undefined, target: string | undefined) => {
  if (target !== '_blank') {
    return rel
  }

  if (!rel) {
    return 'noopener'
  }

  const isRelSafe = rel.includes('noreferrer') || rel.includes('noopener')

  return isRelSafe ? rel : rel.concat(' noopener')
}

export type Props = BaseProps &
  AnchorHTMLAttributes<HTMLAnchorElement> & {
    /** Content of the component */
    children?: ReactNode
    /** Destination the link points to */
    href?: string
    /** Callback invoked when component is clicked */
    onClick?: (event: React.MouseEvent<HTMLAnchorElement>) => void
    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     */
    as?: ElementType
    /** Either it's a regular hyperlink or an _action_ */
    variant?: VariantType
    /** Controls color of the link */
    color?: ColorType
    /** Indicates the order of receiving focus. If not set will not receive focus. */
    tabIndex?: number
    /** Indicates that the user cannot interact with the Link or its children */
    disabled?: boolean

    visited?: boolean
    /**
     * If true, underline decoration never applies
     */
    noUnderline?: boolean
    'aria-disabled'?: boolean
  }

export const Link: OverridableComponent<Props> = forwardRef<
  HTMLAnchorElement,
  Props
>(function Link(props, ref) {
  const {
    href,
    onClick,
    children,
    className,
    color = 'blue',
    style,
    as = 'a',
    variant = 'anchor',
    tabIndex,
    target,
    rel,
    disabled,
    visited = false,
    noUnderline,
    'aria-disabled': ariaDisabled,
    ...rest
  } = props
  const nativeHTMLAttributes = rest
  const classes = useStyles()
  const sanitizedRel = sanitizeRel(rel, target)

  return (
    <MUILink
      {...nativeHTMLAttributes}
      ref={ref}
      href={disabled ? undefined : href}
      target={disabled ? undefined : target}
      rel={sanitizedRel}
      onClick={disabled ? undefined : onClick}
      className={cx(classes.root, className, {
        [classes.action]: variant === 'action',
        [classes.white]: color === 'white',
        [classes.blue]: color === 'blue',
        [classes.visited]: visited,
        [classes.disabled]: disabled,
        [classes.noUnderline]: noUnderline,
      })}
      style={style}
      component={as}
      tabIndex={tabIndex}
      aria-disabled={disabled || ariaDisabled}
    >
      {children}
    </MUILink>
  )
})

Link.defaultProps = {
  as: 'a',
  color: 'blue',
  variant: 'anchor',
  noUnderline: false,
}

Link.displayName = 'Link'

export default Link
