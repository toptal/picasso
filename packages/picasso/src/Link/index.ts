import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Link'

export { default } from './Link'
export type LinkProps = OmitInternalProps<Props>
