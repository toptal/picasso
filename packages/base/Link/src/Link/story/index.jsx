import { Link } from '../Link'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Components').createPage(
  'Link',
  `
    The Link component allows you to easily customize anchor elements with your theme colors and typography styles.

    ${PicassoBook.createBaseDocsLink(
      'https://app.abstract.com/projects/1b06c884-06af-482a-bf12-a82f521a19a1/branches/master/commits/c1f02ee345bc43e9617f5339aaa3eff1b2501d85/files/e67e971a-851c-4375-9821-686c7d4c57e6/layers/BA5476AB-47D9-414D-9FCA-59021BC47C92?mode=design&present=true&sha=c1f02ee345bc43e9617f5339aaa3eff1b2501d85'
    )}

    ${PicassoBook.createSourceLink(__filename)}
  `
)

page.createTabChapter('Props').addComponentDocs({
  component: Link,
  name: 'Link',
  // react-docgen-typescript doesn't work with generics, so we have to generate this
  // documentation manually
  additionalDocs: {
    children: {
      name: 'children',
      type: 'ReactNode',
      description: 'Content of the component',
    },
    href: {
      name: 'href',
      type: 'string',
      description: 'Destination the link points to',
    },
    onClick: {
      name: 'onClick',
      type: {
        name: 'function',
        description:
          '(event: MouseEvent<HTMLAnchorElement, MouseEvent>) => void',
      },
      description: 'Callback invoked when component is clicked',
    },
    as: {
      name: 'as',
      type: {
        name: 'enum',
        enums: [
          '"symbol"',
          '"object"',
          '"style"',
          '"title"',
          '"a"',
          '"abbr"',
          '"address"',
          '"area"',
          '"article"',
          '"aside"',
          '"audio"',
          '"b"',
          '"base"',
          '"bdi"',
          '"bdo"',
          '"big"',
          '"blockquote"',
          '"body"',
          '"br"',
          '... 157 more ...',
        ],
      },
      defaultValue: 'a',
      description:
        'The component used for the root node.\nEither a string to use a DOM element or a component.',
    },
    variant: {
      name: 'variant',
      type: {
        name: 'enum',
        enums: ['"default"', '"action"'],
      },
      defaultValue: 'default',
      description: "Either it's a regular link or an _action_.",
    },
    color: {
      name: 'color',
      type: {
        name: 'enum',
        enums: ['"blue"', '"white"'],
      },
      defaultValue: 'blue',
      description: 'Controls color of the link',
    },
    className: {
      name: 'className',
      type: 'string',
      description: 'Classnames applied to root element',
    },
    style: {
      name: 'style',
      type: 'CSSProperties',
      description: 'Style applied to root element',
    },
    size: {
      name: 'size',
      type: {
        name: 'enum',
        enums: ['"default"', '"inherit"'],
      },
      defaultValue: 'default',
      description:
        'Controls font size of Link, whenever you are using Link inside another text with a different font size you should inherit parents font size.',
    },
  },
})

page
  .createChapter()
  .addExample('Link/story/Default.example.tsx', 'Default')
  .addExample('Link/story/Action.example.tsx', 'Action')
  .addExample('Link/story/FontSize.example.tsx', {
    title: 'Font size',
    description:
      "You can modify Link's font size by wrapping it in `Typography`",
  })
  .addExample('Link/story/Color.example.tsx', 'Color')
  .addExample('Link/story/Disabled.example.tsx', 'Disabled')
  .addExample('Link/story/Routing.example.tsx', {
    title: 'Routing',
    description:
      'An example how to use `@toptal/picasso` Link with `react-router-dom` Link',
    takeScreenshot: false,
  })
