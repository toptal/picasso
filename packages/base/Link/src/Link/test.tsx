import React from 'react'
import { render, fireEvent } from '@toptal/picasso-test-utils'
import { MemoryRouter, Link as RouterLink } from 'react-router-dom'

import { Link } from '../Link'

describe('Link', () => {
  it('renders', () => {
    const { container } = render(<Link>Please verify your email</Link>)

    expect(container).toMatchSnapshot()
  })

  it('renders native attributes', () => {
    const { container } = render(
      <Link
        onBlur={() => window.alert('onBlur')}
        rel='noopener'
        target='_blank'
        download='filename'
        href='https://toptal.com/filename.txt'
      >
        Please verify your email
      </Link>
    )

    expect(container).toMatchSnapshot()
  })

  it('renders a Link from react-router', () => {
    const { container } = render(
      <MemoryRouter>
        <div>
          <Link as={RouterLink} to='/'>
            Please verify your email
          </Link>
        </div>
      </MemoryRouter>
    )

    expect(container).toMatchSnapshot()
  })

  it('renders disabled link', () => {
    const { container } = render(
      <Link
        rel='noopener'
        target='_blank'
        download='filename'
        href='https://toptal.com/filename.txt'
        disabled
      >
        Please verify your email
      </Link>
    )

    expect(container).toMatchSnapshot()
  })

  it('adds rel="noopener" to target="_blank" links in its absence', () => {
    render(
      <Link href='http://example.com' target='_blank'>
        External Link
      </Link>
    )

    expect(document.querySelector('a')).toHaveAttribute('rel', 'noopener')
  })

  it('does not add rel="noopener" to target="_blank" if noreferrer is present', () => {
    render(
      <Link href='http://example.com' target='_blank' rel='noreferrer'>
        External Link
      </Link>
    )

    expect(document.querySelector('a')).not.toHaveAttribute('rel', 'noopener')
  })

  it('does not allow onClick when disabled', () => {
    const onClick = jest.fn()
    const { getByTestId } = render(
      <Link data-testid='foo' onClick={onClick} href='https://foo.bar' disabled>
        Test
      </Link>
    )

    fireEvent.click(getByTestId('foo'))
    expect(onClick).not.toHaveBeenCalled()
  })

  it('does not have href when disabled', () => {
    const { getByTestId } = render(
      <Link data-testid='foo' href='https://foo.bar' disabled>
        Test
      </Link>
    )

    expect(getByTestId('foo')).not.toHaveAttribute('href')
  })
})
