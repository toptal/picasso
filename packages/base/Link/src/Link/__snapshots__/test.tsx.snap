// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <a
      class="MuiTypography-root MuiLink-root MuiLink-underlineHover PicassoLink-root PicassoLink-blue MuiTypography-colorPrimary"
    >
      Please verify your email
    </a>
  </div>
</div>
`;

exports[`Link renders a Link from react-router 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div>
      <a
        class="MuiTypography-root MuiLink-root MuiLink-underlineHover PicassoLink-root PicassoLink-blue MuiTypography-colorPrimary"
        href="/"
      >
        Please verify your email
      </a>
    </div>
  </div>
</div>
`;

exports[`Link renders disabled link 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <a
      aria-disabled="true"
      class="MuiTypography-root MuiLink-root MuiLink-underlineHover PicassoLink-root PicassoLink-blue PicassoLink-disabled MuiTypography-colorPrimary"
      download="filename"
      rel="noopener"
    >
      Please verify your email
    </a>
  </div>
</div>
`;

exports[`Link renders native attributes 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <a
      class="MuiTypography-root MuiLink-root MuiLink-underlineHover PicassoLink-root PicassoLink-blue MuiTypography-colorPrimary"
      download="filename"
      href="https://toptal.com/filename.txt"
      rel="noopener"
      target="_blank"
    >
      Please verify your email
    </a>
  </div>
</div>
`;
